File: Best-Time-To-Buy-And-Sell-Stock-2.cpp
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(prices.size()==1){
            return 0;
        }
        int prev = 0;
        int current = 1;
        int sum = 0;
        
        while(current<n){
            if(prices[current]-prices[prev]>0){
                sum = sum + prices[current]-prices[prev];
            }
            current++;
            prev++;
        }
        return sum;
    }
};

File: BestTimetoBuyandSellStock2.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
# Using peak and valley approach
# Iterate over the prices array starting from 1 and check if the current price is greater than the previous price then add difference to profit
# Return the cumulative profit


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return None
        profit = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                profit += prices[i] - prices[i - 1]
        return profit